flight/PiOS/STM32F4xx/pios_dmashot.c

```diff
@@ -0,0 +1,591 @@
+ * @retval TRUE on success, FALSE if the channel's not set up for DMA.
+ */
+void PIOS_DMAShot_WriteValue(const struct pios_tim_channel *servo_channel, uint16_t throttle)
+{
+	// Fail hard if trying to write values without configured DMAShot. We shouldn't be getting to
+	// this point in that case.
+	PIOS_Assert(dmashot_cfg);
+
+	// If we can't find a timer for this, something's wrong. We shouldn't be getting to this point.
+	// Fail hard.
+	struct servo_timer *s_timer = PIOS_DMAShot_GetServoTimer(servo_channel);
+	PIOS_Assert(s_timer);
+
+	int shift = (servo_channel->timer_chan - s_timer->low_channel) >> 2;
+	int channels = PIOS_DMAShot_GetNumChannels(s_timer);
+
+	if (throttle > 2047)
+		throttle = 2047;
+
+	throttle <<= 5;
+	throttle |=
+			((throttle >> 4 ) & 0xf) ^
+			((throttle >> 8 ) & 0xf) ^
+			((throttle >> 12) & 0xf);
+
+	// Leading zero, trailing zero.
+	for (int i = DMASHOT_MESSAGE_PAUSE; i < DMASHOT_MESSAGE_WIDTH+DMASHOT_MESSAGE_PAUSE; i++) {
+		int addr = i * channels + shift;
+		if (PIOS_DMAShot_HalfWord(s_timer)) {
+			s_timer->buffer.hw[addr] = throttle & 0x8000 ? s_timer->duty_cycle_1 : s_timer->duty_cycle_0;
+		} else {
+			s_timer->buffer.fw[addr] = throttle & 0x8000 ? s_timer->duty_cycle_1 : s_timer->duty_cycle_0;
+		}
+		throttle <<= 1;
+	}
+}
+
+/**
+ * @brief Tells the DMAShot driver about a timer that needs to be set up.
+ * @param[in] timer The STM32 timer in question.
+ * @param[in] clockrate The frequency the timer's set up to run.
+ * @param[in] dshot_freq The desired DShot signal frequency (in KHz).
+ * @retval TRUE on success, FALSE when there's no configuration for the specific timer,
+                        or DMAShot isn't set up correctly.
+ */
+bool PIOS_DMAShot_RegisterTimer(TIM_TypeDef *timer, uint32_t clockrate, uint32_t dshot_freq)
+{
+	// No configuration, push for GPIO in upstream.
+	if (!dmashot_cfg || !servo_timers)
+		return false;
+
+	// Check whether the timer is configured for DMA first. If not, bail and
+	// tell upstairs that DMA DShot is a no-go.
+	const struct pios_dmashot_timer_cfg *dma_config = NULL;
+
+	for (int i = 0; i < dmashot_cfg->num_timers; i++) {
+		if (dmashot_cfg->timer_cfg[i].timer == timer) {
+			dma_config = &dmashot_cfg->timer_cfg[i];
+		}
+	}
+
+	if (!dma_config)
+		return false;
+
+	// Check whether there's an existing config from a previous registration,
+	// since we can't clean up due to a lack of free(). Timer/pin stuff is
+	// static anyway.
+	struct servo_timer *s_timer = NULL;
+
+	bool found = false;
+	for (int i = 0; i < MAX_TIMERS; i++) {
+		if (servo_timers[i] && servo_timers[i]->dma->timer == timer) {
+			s_timer = servo_timers[i];
+			found = true;
+			break;
+		}
+	}
+
+	// Nothing found? Allocate a new slot for the timer.
+	if (!found) {
+		for (int i = 0; i < MAX_TIMERS; i++) {
+			if (!servo_timers[i]) {
+				s_timer = PIOS_malloc_no_dma(sizeof(struct servo_timer));
+				memset(s_timer, 0, sizeof(struct servo_timer));
```

Comment:
I'm ignoring allocation / memset for now.. in my next review pass I'll look at them closely as they require a focused review.